### ISSUE TITLE
Converting mean anomaly to true anomaly

### ISSUE THREAD

**Issue by s-m-e** [poliastro/poliastro#1013]
`Orbit.from_classical` "only" accepts a true anomaly. It lacks support for a mean anomaly as an alternative. poliastro also lacks a method for converting from mean anomaly to true anomaly in its [`angle` module](https://github.com/poliastro/poliastro/blob/0.14.x/src/poliastro/core/angles.py).

💡 **Feature request**

A good, working and MIT-licenced implementation for converting from mean to true anomaly can be found [here](https://github.com/RazerM/orbital/blob/73437b627bb2f816612bd4e9c39010483d3c1a29/orbital/utilities.py#L73). It could be adopted and accelerated with numba.

🎯 **Goal**

Virtually every database for solar system objects' orbital elements provides mean anomalies.

**Comment by astrojuanlu** [comment]
> poliastro also lacks a method for converting from mean anomaly to true anomaly in its angle module.

This confinement, mean anomalies have become a favourite topic of mine :smiling_imp: Take a look at this comment:

https://github.com/poliastro/poliastro/pull/908#issuecomment-615889854

In short: the "mean anomaly" is by no means (pun intended) a well defined quantity with a clear geometrical or physical meaning (no pun intended) _for all possible eccentricity values_, and therefore generic conversions should be avoided.

In fact, we _used_ to have `M_to_nu` and `nu_to_M` functions! But I ended up removing them after this realization.

However, if you are analyzing asteroid trajectories and know for real their eccentricity, you can use the well known `M_to_E` and `E_to_nu` that are available in `poliastro.core.angles`, as it's done in these two places in poliastro already:

https://github.com/poliastro/poliastro/blob/3d85e79ea67ec3bf8a575243a7a27de3728940a9/src/poliastro/neos/dastcom5.py#L370-L378

https://github.com/poliastro/poliastro/blob/c516ac476efa00353649077ed63162148cd617ec/src/poliastro/twobody/orbit.py#L683-L691

Let me know what you think.

**Comment by astrojuanlu** [comment]
> Virtually every database for solar system objects' orbital elements provides mean anomalies.

I also observed that, and none of those databases cares to define what's the mean anomaly for the non elliptic case for them, or how near-parabolic cases should be handled (there's a singularity in the vicinity of ecc = 1).

**Comment by s-m-e** [comment]
> In short: the "mean anomaly" is by no means (pun intended) a well defined quantity with a clear geometrical or physical meaning (no pun intended) for all possible eccentricity values, and therefore generic conversions should be avoided.

Wt.. Interesting, literally. I dived into this topic a couple of years ago and thought it was me not understanding the literature.

> you can use the well known M_to_E and E_to_nu that are available in poliastro.core.angles

Will try and report back, thanks. If this code-snipped exists twice in poliastro already, why not put it into a function (i.e. reinstate one of the old conversion functions) and let the function throw a *big fat red warning*? I get it, it sucks, but there are a few serious use cases ...

> I also observed that, and none of those databases cares to define what's the mean anomaly for the non elliptic case for them, or how near-parabolic cases should be handled (there's a singularity in the vicinity of ecc = 1).

Hmm. Sounds like it's worth putting a short & proper text together for contacting some of those folks, like "given the presented/outlined confusing state of astrodynamics, what is *your* interpretation of this property?". It *could* explain a few interesting inconsistencies that I have been observing between different data sources.

**Comment by s-m-e** [comment]
Works. Tested against `orbital` and 1M orbital elements from MPC:

```python
import numpy as np
import astropy.units as u

from poliastro.twobody.angles import (
    D_to_nu, E_to_nu, F_to_nu, M_to_D, M_to_E, M_to_F,
)
from orbital.utilities import true_anomaly_from_mean

def true_anomaly_from_mean_orbital(ecc, M):
    nu = true_anomaly_from_mean(
        e = float(ecc.value),
        M = float(M.to(u.rad).value)
    )
    if nu >= np.pi:
        nu -= 2 * np.pi
    return nu * u.rad

def true_anomaly_from_mean_poliastro(ecc, M):
    if ecc < 1:
        M = (M + np.pi * u.rad) % (2 * np.pi * u.rad) - np.pi * u.rad
        return E_to_nu(M_to_E(M, ecc), ecc).to(u.rad)
    elif ecc == 1: 
        return D_to_nu(M_to_D(M)).to(u.rad)
    else: 
        return F_to_nu(M_to_F(M, ecc), ecc).to(u.rad)
```

Absolute differences in results vary from `0.0` to `4.45e-12`, mean `3.89e-16`, median `3.33e-16`. 

Notably, `orbital`'s pure Python (no numba) implementation is roughly 10 times faster than what I wrote based on `poliastro`. I guess if I strip away a ton of unit conversions and go for the methods in `poliastro.core.angles` directly instead of the methods in `poliastro.twobody.angles`, it should improve the result.

**Comment by s-m-e** [comment]
Jep, about twice as fast as `orbital` (i.e. 20 times as fast as my above implementation):

```python
import numpy as np

from poliastro.core.angles import (
    D_to_nu as D_to_nu_fast,
    E_to_nu as E_to_nu_fast,
    F_to_nu as F_to_nu_fast,
    M_to_D as M_to_D_fast,
    M_to_E as M_to_E_fast,
    M_to_F as M_to_F_fast,
)

def true_anomaly_from_mean_poliastro_fast(ecc, M):
    ecc, M = ecc.value, M.to(u.rad).value
    if ecc < 1:
        M = (M + np.pi) % (2 * np.pi) - np.pi
        return E_to_nu_fast(M_to_E_fast(M, ecc), ecc) * u.rad
    elif ecc == 1: 
        return D_to_nu_fast(M_to_D_fast(M)) * u.rad
    else: 
        return F_to_nu_fast(M_to_F_fast(M, ecc), ecc) * u.rad
```

Statistics (compared to `orbital`) changed a little. Differences from 0.0 to 4.76e-12, mean 3.57e-16, median 2.22e-16.

**Comment by astrojuanlu** [comment]
Yeah, it's crazy how slow unit conversions are. But that's why we have two layers :)

Regarding the option of providing a function with a big warning... I prefer not to, as a way of telling the user "if it's difficult, it's a signal you're doing something wrong". Conversely, that forces us making correct operations easy. Also, people never read the warnings :D

**Comment by astrojuanlu** [comment]
I think we can close this, feel free to complain here otherwise @s-m-e !

**Comment by s-m-e** [comment]
@astrojuanlu I'd suggest to at least put a dedicated section into your documentation, explaining your rational of not having such a function, the state of the art in astrodynamics and possible "workarounds". Otherwise people have to collect this information from the issue tracker ...

I am also still thinking that this could be an excellent topic for a fun & controversial, short two to three pages review paper.

**Comment by astrojuanlu** [comment]
Good point, reopening and tagging accordingly.

> I am also still thinking that this could be an excellent topic for a fun & controversial, short two to three pages review paper.

Hmmm tell me more... 😃

**Comment by s-m-e** [comment]
> Hmmm tell me more... 

I can primarily speak for the small body science crowd ... how about something along the lines of "There is no mean anomaly!" ... submitted to [Icarus](https://www.journals.elsevier.com/icarus) or some place similar ... The audience will be cheering, I guarantee ;)

**Comment by s-m-e** [comment]
@astrojuanlu If your following statement is actually true, ...

> In short: the "mean anomaly" is by no means (pun intended) a well defined quantity with a clear geometrical or physical meaning (no pun intended) for all possible eccentricity values, and therefore generic conversions should be avoided.

... can mathematically be proven, geometrically shown and cemented by a few possibly contradictory literature references, this may have at least a bit of an impact (in this field and others). I bet that there aren't many people who are catching those "[inconsistencies](https://github.com/poliastro/poliastro/pull/908#issuecomment-615889854)" and/or act upon them the "right" way. One can go with a less provocative title, but the bottom line is that this is a bit of a dormant problem I suppose.

**Comment by astrojuanlu** [comment]
Coming back to this (the docs part, not the scholar article part)

According to the documentation classification we reviewed in #975, I think this would qualify as a "Discussion/Explanation". So we just have to find a good name for them (I am not quite sold on these) and write the first one. The idea is that these Explanations should focus more on the theoretical aspects, and use the code as a support. In fact, I believe some of our notebooks could be moved to this new category.

What about "In depth"?

**Comment by s-m-e** [comment]
> What about "In depth"?

Informal way of saying "discussion", is not it ;) ? Sounds good.

**Comment by astrojuanlu** [comment]
The definition of the mean anomaly is (the Kepler equation)

$$
M = E - e \\, \sin E
$$

where $E$ is the eccentric anomaly and $e$ is the eccentricity of the orbit.

**Comment by astrojuanlu** [comment]
(Just testing the new equations feature 😍)
